File: PYTHON/mesh.py
import numpy as np


class Mesh:
    def __init__(self):
        self.meshtype = None
        self.subdiv = 0
        self.vertices = []
        self.indices = []

    def vertex_count(self):
        return len(self.vertices)

    def triangles_count(self):
        return len(self.indices) // 3


def load_cube(mesh, subdiv):
    mesh.meshtype = "CUBE"
    mesh.subdiv = subdiv

    n = subdiv + 1

    faces = [
        {"axis": 0, "dir": 0},
        {"axis": 1, "dir": 0},
        {"axis": 2, "dir": 0},
        {"axis": 0, "dir": 1},
        {"axis": 1, "dir": 1},
        {"axis": 2, "dir": 1},
    ]

    total_vertices = 6 * n * n
    unique_vertices = 6 * subdiv * subdiv + 2

    # 使用一个字典模拟哈希表
    vertex_index_map = {}
    mesh.vertices = [None] * unique_vertices

    dup_to_nodup_index = np.zeros(total_vertices, dtype=int)

    t = 0  # 表示存在重复点的下标
    p = 0  # 表示不重复点的下标

    inv_subdiv = 1.0 / subdiv

    for face in faces:
        axis = face["axis"]
        dir = face["dir"]

        idx1 = (axis + 1) % 3
        idx2 = (axis + 2) % 3

        for i in range(n):
            for j in range(n):
                coords = [0, 0, 0]
                coords[axis] = dir * subdiv
                coords[idx1] = i
                coords[idx2] = j

                # 计算哈希值
                key = coords[0] | (coords[1] << 20) | (coords[2] << 40)

                if key not in vertex_index_map:  # 如果是首次出现
                    vertex_index_map[key] = p
                    dup_to_nodup_index[t] = p

                    # 计算顶点位置
                    fx = coords[0] * inv_subdiv * 2.0 - 1.0
                    fy = coords[1] * inv_subdiv * 2.0 - 1.0
                    fz = coords[2] * inv_subdiv * 2.0 - 1.0

                    mesh.vertices[p] = (fx, fy, fz)

                    p += 1
                else:
                    dup_to_nodup_index[t] = vertex_index_map[key]

                t += 1

    mesh.indices = [0] * (36 * subdiv * subdiv)
    t = 0
    face_vertex_offset = 0

    for face_idx in range(6):
        for i in range(subdiv):
            for j in range(subdiv):
                idx0 = face_vertex_offset + i * n + j
                idx1 = face_vertex_offset + i * n + j + 1
                idx2 = face_vertex_offset + (i + 1) * n + j
                idx3 = face_vertex_offset + (i + 1) * n + j + 1

                v0 = dup_to_nodup_index[idx0]
                v1 = dup_to_nodup_index[idx1]
                v2 = dup_to_nodup_index[idx2]
                v3 = dup_to_nodup_index[idx3]

                mesh.indices[t] = v0
                mesh.indices[t + 1] = v1
                mesh.indices[t + 2] = v2
                mesh.indices[t + 3] = v1
                mesh.indices[t + 4] = v3
                mesh.indices[t + 5] = v2
                t += 6

        face_vertex_offset += n * n

    return 0


class CSRMatrix:
    def __init__(self, mesh):
        """
        根据网格初始化CSR矩阵
        """
        self.mesh = mesh
        self.rows = len(mesh.vertices)
        self.cols = self.rows

        # 初始化 row_offset
        self.row_offset = np.ones(self.rows + 1, dtype=int)

        # 统计每个顶点的出现次数
        for t in range(mesh.triangle_count()):
            self.row_offset[mesh.indices[3 * t + 0]] += 1
            self.row_offset[mesh.indices[3 * t + 1]] += 1
            self.row_offset[mesh.indices[3 * t + 2]] += 1

        # 将数量转换为偏移量
        for i in range(1, self.rows):
            self.row_offset[i + 1] += self.row_offset[i]

        # 向后移动一位
        self.row_offset[1:] = self.row_offset[:-1]
        self.row_offset[0] = 0

        total_nonzeros = self.row_offset[self.rows]  # 最后一个元素为非零元素总数
        self.elements = np.zeros(total_nonzeros)  # 非零元素数组
        self.elm_idx = -np.ones(total_nonzeros, dtype=int)  # 列索引数组，初始化为 -1

        # 填充非零元素索引
        for t in range(mesh.triangle_count()):
            a, b, c = mesh.indices[3 * t : 3 * t + 3]
            triangle = [a, b, c]

            for current_vtx in triangle:
                for current_row in triangle:
                    offset = self.row_offset[current_row]
                    length = self.row_offset[current_row + 1] - offset

                    for i in range(length):
                        if self.elm_idx[offset + i] == current_vtx:
                            break  # 已经出现过，跳过
                        elif self.elm_idx[offset + i] == -1:  # 第一次出现，插入
                            self.elm_idx[offset + i] = current_vtx
                            break

        # 按大小排序每行的非零元素索引
        for row in range(self.rows):
            offset = self.row_offset[row]
            length = self.row_offset[row + 1] - offset
            self.elm_idx[offset : offset + length] = np.sort(
                self.elm_idx[offset : offset + length]
            )

    def mvp(self, x, y):
        """
        稀疏矩阵乘向量 (Matrix-Vector Product)
        """
        if len(x) != self.cols or len(y) != self.rows:
            raise ValueError(
                "Size mismatch: The number of columns in the matrix does not match the size of the vector."
            )

        y.fill(0.0)  # 初始化结果向量

        for r in range(self.rows):
            offset = self.row_offset[r]
            length = self.row_offset[r + 1] - offset

            local_sum = 0.0
            for i in range(length):
                col = self.elm_idx[offset + i]
                local_sum += self.elements[offset + i] * x[col]

            y[r] += local_sum


from collections import defaultdict


def get_Mloc_value(i, j, Mloc):
    """
    获取局部质量矩阵中的值
    """
    return Mloc[0] if i == j else Mloc[1]


def build_vertex_to_local_index(a, b, c):
    """
    构建顶点到局部索引的映射
    """
    return {a: 0, b: 1, c: 2}


def process_mass_matrix_row(M, current_row, i, vertex_to_local_index, Mloc):
    """
    处理质量矩阵中的一行
    """
    offset = M.row_offset[current_row]
    length = M.row_offset[current_row + 1] - offset

    for idx in range(length):
        col = M.elm_idx[offset + idx]

        # 检查列是否在当前三角形的顶点中
        if col in vertex_to_local_index:
            j = vertex_to_local_index[col]
            value = get_Mloc_value(i, j, Mloc)

            # 原子加法，避免并行冲突
            M.elements[offset + idx] += value


def build_mass_matrix(M):
    """
    构建全局质量矩阵
    """
    mesh = M.mesh

    for t in range(mesh.triangle_count()):
        # 获取三角形顶点
        a, b, c = mesh.indices[3 * t : 3 * t + 3]

        # 获取顶点坐标
        A = mesh.vertices[a]
        B = mesh.vertices[b]
        C = mesh.vertices[c]

        # 计算 AB 和 AC
        AB = B - A
        AC = C - A

        # 计算局部质量矩阵
        Mloc = mass_loc(AB, AC)

        # 构建顶点到局部索引的映射
        vertex_to_local_index = build_vertex_to_local_index(a, b, c)

        # 遍历三角形的每一个顶点（对应全局矩阵的行）
        for i, current_row in enumerate([a, b, c]):
            process_mass_matrix_row(M, current_row, i, vertex_to_local_index, Mloc)


def mass_loc(AB, AC):
    """
    计算局部质量矩阵的两个值
    """
    cross_product = np.cross(AB, AC)
    S_ABC = 0.5 * np.linalg.norm(cross_product)
    return S_ABC / 6.0, S_ABC / 12.0

File: lib/src/Mesh/Mesh.cpp
#include <Mesh.h>
#include <TArray.h>
#include <vec3.h>
#include <cstdint>
// #include <timer.h>
// #include <iostream>
#include <unordered_map>

/* 生成立方体网格, 中心为原点，边长为2
 * 对于有n个子分割的网格
 * 正方体网格中有6n^2+2个顶点, 12n^2个三角形
 * 1. 生成顶点，同时使用hash表记录每个顶点的对应下标，生成一个不包含重复点的顶点集合 vertex_index_map
 * 2. 同时生成一个有重复点时下标与无重复点时下标的hash表 dupToNoDupIndex
 * 3. 根据dupToNoDupIndex创建三角形
 */

int load_cube(Mesh &m, const int subdiv)
{
    m.meshtype = CUBE;
    m.subdiv = subdiv;
    // Timer timer;
    // timer.start();

    int n = subdiv + 1;

    struct Face
    {
        int axis;      // 0:x, 1:y, 2:z
        int dir;       // 0:负方向, 1:正方向
        int firstAxis; // 当前面进行编号的坐标轴顺序，以保持三角形的方向性
        int lastAxis;
    };

    std::vector<Face> faces = {{0, 1, 1, 2},
                               {1, 1, 0, 2},
                               {0, 0, 1, 2},
                               {1, 0, 0, 2},
                               {2, 1, 1, 0},
                               {2, 0, 1, 0}};

    int totalVertices = 6 * n * n;                // 重复顶点有6n^2个
    int uniqueVertices = 6 * subdiv * subdiv + 2; // 不重复的顶点有6 * subdiv^2 + 2个

    // 使用一个hash函数将Vec3映射到int64, 避免使用Vec3作为主键
    std::unordered_map<int64_t, int> vertex_index_map;
    vertex_index_map.reserve(uniqueVertices);
    m.vertices.resize(uniqueVertices);

    // 从重复点索引到不重复点索引的map
    int *dupToNoDupIndex = new int[totalVertices];

    int t = 0; // t 表示存在重复点的下标
    int p = 0; // p 表示不重复点的下标

    double invSubdiv = 1.0 / (double)subdiv; // 提前计算好减少浮点数除法

    for (const Face &face : faces)
    {
        int axis = face.axis;
        int dir = face.dir;

        int idx1 = face.firstAxis;
        int idx2 = face.lastAxis;

        for (int i = 0; i < n; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                /* 先使用整数坐标表示一个点，方便后续使用hash
                 * 例如face = {0, 1}, 即表示垂直于x轴，位于正半轴区域的面
                 * axis = 0, dir = 1, idx1 = 1, idx2 = 2
                 * 经过处理后coords存储的是 [subdiv, i, j] 这样的点
                 * 之后经过处理可以变成需要的范围
                 */
                int coords[3] = {0, 0, 0};
                coords[axis] = dir * subdiv;

                coords[idx1] = j; // 此处firstAxis上的先编号
                coords[idx2] = i;

                /* Hash函数的部分
                 * 将三个坐标映射到int64的范围中作为主键
                 * 需要subdiv < 2^20
                 */
                int64_t key = (int64_t)coords[0] | (int64_t)coords[1] << 20 | (int64_t)coords[2] << 40;

                auto it = vertex_index_map.find(key);
                if (it == vertex_index_map.end()) // 若这个点是首次出现
                {
                    vertex_index_map[key] = p; // 存储这个点的位置为p
                    dupToNoDupIndex[t] = p;

                    // 计算顶点的位置
                    double fx = coords[0] * invSubdiv * 2.0 - 1.0;
                    double fy = coords[1] * invSubdiv * 2.0 - 1.0;
                    double fz = coords[2] * invSubdiv * 2.0 - 1.0;

                    m.vertices[p] = {fx, fy, fz};

                    ++p;
                }
                else
                {
                    dupToNoDupIndex[t] = it->second; // 将重复点的下标对应到不重复点的下标
                }

                ++t;
            }
        }
    }
    // timer.stop();

    // std::cout << "生成网格用时: " << timer.elapsedMilliseconds() << "ms\n";

    // timer.start();
    //  生成三角形
    m.indices.resize(36 * subdiv * subdiv); // 共计12n^2个三角形, 36n^2个顶点

    t = 0;
    int faceVertexOffset = 0;
    for (int faceIdx = 0; faceIdx < 6; ++faceIdx)
    {
        for (int i = 0; i < subdiv; ++i)
        {
            for (int j = 0; j < subdiv; ++j)
            {
                int idx0 = faceVertexOffset + i * n + j;           // 左下的点
                int idx1 = faceVertexOffset + i * n + j + 1;       // 右下
                int idx2 = faceVertexOffset + (i + 1) * n + j;     // 左上
                int idx3 = faceVertexOffset + (i + 1) * n + j + 1; // 右上

                int v0 = dupToNoDupIndex[idx0];
                int v1 = dupToNoDupIndex[idx1];
                int v2 = dupToNoDupIndex[idx2];
                int v3 = dupToNoDupIndex[idx3];

                if (faceIdx == 1 || faceIdx == 2 || faceIdx == 4)
                {
                    m.indices[t++] = v1;
                    m.indices[t++] = v0;
                    m.indices[t++] = v3;
                    m.indices[t++] = v0;
                    m.indices[t++] = v2;
                    m.indices[t++] = v3;
                }
                else
                {
                    m.indices[t++] = v0;
                    m.indices[t++] = v1;
                    m.indices[t++] = v2;
                    m.indices[t++] = v1;
                    m.indices[t++] = v3;
                    m.indices[t++] = v2;
                }
            }
        }

        faceVertexOffset += n * n;
    }

    delete[] dupToNoDupIndex;
    // timer.stop();

    // std::cout << "生成三角形用时: " << timer.elapsedMilliseconds() << "ms\n";

    return 0;
}

int load_sphere(Mesh &m, int subdiv)
{
    load_cube(m, subdiv);
    m.meshtype = SPHERE;

    // Timer t;
    // t.start();
    for (size_t i = 0; i < m.vertex_count(); ++i)
    {
        m.vertices[i] = normalized(m.vertices[i]);
    }
    // t.stop();
    // std::cout << "生成球面用时: " << t.elapsedMilliseconds() << "ms" << std::endl;
    return 0;
}

int load_cube(Mesh &m, const int subdiv, int *dupToNoDupIndex)
{
    m.meshtype = CUBE;
    m.subdiv = subdiv;
    // Timer timer;
    // timer.start();

    int n = subdiv + 1;

    struct Face
    {
        int axis;      // 0:x, 1:y, 2:z
        int dir;       // 0:负方向, 1:正方向
        int firstAxis; // 当前面进行编号的坐标轴顺序，以保持三角形的方向性
        int lastAxis;
    };

    std::vector<Face> faces = {{0, 1, 1, 2},
                               {1, 1, 0, 2},
                               {0, 0, 1, 2},
                               {1, 0, 0, 2},
                               {2, 1, 1, 0},
                               {2, 0, 1, 0}};

    int totalVertices = 6 * n * n;                // 重复顶点有6n^2个
    int uniqueVertices = 6 * subdiv * subdiv + 2; // 不重复的顶点有6 * subdiv^2 + 2个

    // 使用一个hash函数将Vec3映射到int64, 避免使用Vec3作为主键
    std::unordered_map<int64_t, int> vertex_index_map;
    vertex_index_map.reserve(uniqueVertices);
    m.vertices.resize(uniqueVertices);

    int t = 0; // t 表示存在重复点的下标
    int p = 0; // p 表示不重复点的下标

    double invSubdiv = 1.0 / (double)subdiv; // 提前计算好减少浮点数除法

    for (const Face &face : faces)
    {
        int axis = face.axis;
        int dir = face.dir;

        int idx1 = face.firstAxis;
        int idx2 = face.lastAxis;

        for (int i = 0; i < n; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                /* 先使用整数坐标表示一个点，方便后续使用hash
                 * 例如face = {0, 1}, 即表示垂直于x轴，位于正半轴区域的面
                 * axis = 0, dir = 1, idx1 = 1, idx2 = 2
                 * 经过处理后coords存储的是 [subdiv, i, j] 这样的点
                 * 之后经过处理可以变成需要的范围
                 */
                int coords[3] = {0, 0, 0};
                coords[axis] = dir * subdiv;

                coords[idx1] = j;
                coords[idx2] = i;

                /* Hash函数的部分
                 * 将三个坐标映射到int64的范围中作为主键
                 * 需要subdiv < 2^20
                 */
                int64_t key = (int64_t)coords[0] | (int64_t)coords[1] << 20 | (int64_t)coords[2] << 40;

                auto it = vertex_index_map.find(key);
                if (it == vertex_index_map.end()) // 若这个点是首次出现
                {
                    vertex_index_map[key] = p; // 存储这个点的位置为p
                    dupToNoDupIndex[t] = p;

                    // 计算顶点的位置
                    double fx = coords[0] * invSubdiv * 2.0 - 1.0;
                    double fy = coords[1] * invSubdiv * 2.0 - 1.0;
                    double fz = coords[2] * invSubdiv * 2.0 - 1.0;

                    m.vertices[p] = {fx, fy, fz};

                    ++p;
                }
                else
                {
                    dupToNoDupIndex[t] = it->second; // 将重复点的下标对应到不重复点的下标
                }

                ++t;
            }
        }
    }
    // timer.stop();

    // std::cout << "生成网格用时: " << timer.elapsedMilliseconds() << "ms\n";

    // timer.start();
    //  生成三角形
    m.indices.resize(36 * subdiv * subdiv); // 共计12n^2个三角形, 36n^2个顶点

    t = 0;
    int faceVertexOffset = 0;
    for (int faceIdx = 0; faceIdx < 6; ++faceIdx)
    {
        for (int i = 0; i < subdiv; ++i)
        {
            for (int j = 0; j < subdiv; ++j)
            {
                int idx0 = faceVertexOffset + i * n + j;           // 左下的点
                int idx1 = faceVertexOffset + i * n + j + 1;       // 右下
                int idx2 = faceVertexOffset + (i + 1) * n + j;     // 左上
                int idx3 = faceVertexOffset + (i + 1) * n + j + 1; // 右上

                int v0 = dupToNoDupIndex[idx0];
                int v1 = dupToNoDupIndex[idx1];
                int v2 = dupToNoDupIndex[idx2];
                int v3 = dupToNoDupIndex[idx3];

                if (faceIdx == 1 || faceIdx == 2 || faceIdx == 4)
                {
                    m.indices[t++] = v1;
                    m.indices[t++] = v0;
                    m.indices[t++] = v3;
                    m.indices[t++] = v0;
                    m.indices[t++] = v2;
                    m.indices[t++] = v3;
                }
                else
                {
                    m.indices[t++] = v0;
                    m.indices[t++] = v1;
                    m.indices[t++] = v2;
                    m.indices[t++] = v1;
                    m.indices[t++] = v3;
                    m.indices[t++] = v2;
                }
            }
        }

        faceVertexOffset += n * n;
    }
    // timer.stop();

    // std::cout << "生成三角形用时: " << timer.elapsedMilliseconds() << "ms\n";

    return 0;
}

int load_sphere(Mesh &m, int subdiv, int *dupToNoDupIndex)
{
    load_cube(m, subdiv, dupToNoDupIndex);
    m.meshtype = SPHERE;

    // Timer t;
    // t.start();
    for (size_t i = 0; i < m.vertex_count(); ++i)
    {
        m.vertices[i] = normalized(m.vertices[i]);
    }
    // t.stop();
    // std::cout << "生成球面用时: " << t.elapsedMilliseconds() << "ms" << std::endl;
    return 0;
}

Mesh::Mesh(int subdiv, MeshType meshtype)
    : dupToNoDupIndex(nullptr)
{
    if (meshtype == CUBE)
    {
        load_cube(*this, subdiv);
    }
    else if (meshtype == SPHERE)
    {
        load_sphere(*this, subdiv);
    }
}

Mesh::Mesh(int subdiv, MeshType meshtype, bool saveDTND)
    : dupToNoDupIndex(new int[6 * (subdiv + 1) * (subdiv + 1)])
{
    if (!saveDTND)
    {
        delete[] dupToNoDupIndex;
        dupToNoDupIndex = nullptr;
        if (meshtype == CUBE)
        {
            load_cube(*this, subdiv);
        }
        else if (meshtype == SPHERE)
        {
            load_sphere(*this, subdiv);
        }
    }
    else
    {
        if (meshtype == CUBE)
        {
            load_cube(*this, subdiv, dupToNoDupIndex);
        }
        else if (meshtype == SPHERE)
        {
            load_sphere(*this, subdiv, dupToNoDupIndex);
        }
    }
}

Mesh::~Mesh()
{
    if (dupToNoDupIndex)
    {
        delete[] dupToNoDupIndex;
        dupToNoDupIndex = nullptr;
    }
}